File: impl/csharp/Graph.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System.Collections.Generic;

namespace Naalp
{
    /// <summary>
    /// N-AALP causal graph (C7, §8.2-§8.3) and federation reconcile (T13, §8.6) for the C# SDK.
    ///
    /// <para>VerifyCausal enforces no-future-cause (a present cause may not sit at a later position than
    /// its effect) and acyclicity. Reconcile is the deterministic merge: a topological linearization of
    /// the union causal DAG, ties broken by content id (bytewise), scope-independent (R-8.6).</para>
    /// </summary>
    public static class Graph
    {
        /// <summary>A causal graph node: id bytes, its cause ids, and its position.</summary>
        public sealed class Node
        {
            public readonly byte[] Id;
            public readonly List<byte[]> Causes;
            public readonly long Position;
            public Node(byte[] id, List<byte[]> causes, long position)
            {
                Id = id;
                Causes = causes;
                Position = position;
            }
        }

        private static string Key(byte[] b) => Hex.Encode(b);

        public static void VerifyCausal(List<Node> nodes)
        {
            var idx = new Dictionary<string, int>();
            for (int i = 0; i < nodes.Count; i++)
            {
                idx[Key(nodes[i].Id)] = i;
            }

            // no future cause
            foreach (Node node in nodes)
            {
                foreach (byte[] c in node.Causes)
                {
                    if (idx.TryGetValue(Key(c), out int j) && nodes[j].Position > node.Position)
                    {
                        throw new NaalpException("CausalViolation", "cause at a later position than its effect");
                    }
                }
            }

            // acyclic (3-colour DFS over effect -> cause edges)
            const int White = 0, Gray = 1, Black = 2;
            var color = new int[nodes.Count];

            bool HasCycle(int i)
            {
                color[i] = Gray;
                foreach (byte[] c in nodes[i].Causes)
                {
                    if (!idx.TryGetValue(Key(c), out int j))
                    {
                        continue;
                    }
                    if (color[j] == Gray)
                    {
                        return true;
                    }
                    if (color[j] == White && HasCycle(j))
                    {
                        return true;
                    }
                }
                color[i] = Black;
                return false;
            }

            for (int i = 0; i < nodes.Count; i++)
            {
                if (color[i] == White && HasCycle(i))
                {
                    throw new NaalpException("CausalViolation", "causal graph contains a cycle");
                }
            }
        }

        /// <summary>Deterministic topological merge over the union causal DAG; ties break by content id.</summary>
        public static List<byte[]> Reconcile(List<Node> nodes)
        {
            VerifyCausal(nodes);
            int n = nodes.Count;
            var ids = new byte[n][];
            var present = new HashSet<string>();
            for (int i = 0; i < n; i++)
            {
                ids[i] = nodes[i].Id;
                present.Add(Key(nodes[i].Id));
            }

            // causes filtered to present nodes, as membership sets
            var causes = new List<HashSet<string>>(n);
            var indeg = new int[n];
            for (int i = 0; i < n; i++)
            {
                var set = new HashSet<string>();
                foreach (byte[] c in nodes[i].Causes)
                {
                    string ck = Key(c);
                    if (present.Contains(ck))
                    {
                        set.Add(ck);
                    }
                }
                causes.Add(set);
                indeg[i] = set.Count;
            }

            var done = new bool[n];
            var order = new List<byte[]>(n);
            while (order.Count < n)
            {
                int pick = -1;
                for (int i = 0; i < n; i++)
                {
                    if (done[i] || indeg[i] != 0)
                    {
                        continue;
                    }
                    if (pick == -1 || Cbor.CompareBytes(ids[i], ids[pick]) < 0)
                    {
                        pick = i;
                    }
                }
                if (pick == -1)
                {
                    throw new NaalpException("CausalViolation", "no ready node (unreachable after verify_causal)");
                }
                done[pick] = true;
                order.Add(ids[pick]);
                string pickKey = Key(ids[pick]);
                for (int j = 0; j < n; j++)
                {
                    if (!done[j] && causes[j].Contains(pickKey))
                    {
                        indeg[j] -= 1;
                    }
                }
            }
            return order;
        }

        /// <summary>The tier-1 Reconcile body {1: [authorities], 2: [order content-ids]}.</summary>
        public static byte[] ReconcileRecord(List<string> authorities, List<byte[]> order)
        {
            var auth = new List<Cbor.Value>(authorities.Count);
            foreach (string a in authorities)
            {
                auth.Add(new Cbor.T(a));
            }
            var ordr = new List<Cbor.Value>(order.Count);
            foreach (byte[] o in order)
            {
                ordr.Add(new Cbor.B(o));
            }
            return Cbor.Encode(new Cbor.M(new List<Cbor.Pair>
            {
                new Cbor.Pair(new Cbor.U(1), new Cbor.A(auth)),
                new Cbor.Pair(new Cbor.U(2), new Cbor.A(ordr)),
            }));
        }
    }
}

File: impl/php/src/Graph.php
<?php
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
/**
 * N-AALP causal graph (C7, §8.2-§8.3) and federation reconcile (T13, §8.6) for the PHP SDK.
 *
 * verifyCausal enforces no-future-cause (a present cause may not sit at a later position than its
 * effect) and acyclicity. reconcile is the deterministic merge: a topological linearization of the
 * union causal DAG, ties broken by content id (bytewise), scope-independent (R-8.6).
 *
 * A node is [id: binary string, causes: string[], position: int]. Content ids are used as map
 * keys in hex form to stay clear of PHP's numeric-string key coercion.
 */

declare(strict_types=1);

namespace Naalp;

class CausalViolation extends \RuntimeException
{
    public string $kind = "CausalViolation";
}

final class Graph
{
    /**
     * @param array<int,array{0:string,1:array<int,string>,2:int}> $nodes
     */
    public static function verifyCausal(array $nodes): void
    {
        $idx = [];
        foreach ($nodes as $i => $node) {
            $idx[\bin2hex($node[0])] = $i;
        }
        // no future cause
        foreach ($nodes as $node) {
            [$nid, $causes, $pos] = $node;
            foreach ($causes as $c) {
                $key = \bin2hex($c);
                if (\array_key_exists($key, $idx)) {
                    $j = $idx[$key];
                    if ($nodes[$j][2] > $pos) {
                        throw new CausalViolation("cause at a later position than its effect");
                    }
                }
            }
        }
        // acyclic (3-colour DFS over effect -> cause edges)
        $count = \count($nodes);
        $color = \array_fill(0, $count, 0); // 0 WHITE, 1 GRAY, 2 BLACK

        $hasCycle = function (int $i) use (&$hasCycle, &$color, $nodes, $idx): bool {
            $color[$i] = 1;
            foreach ($nodes[$i][1] as $c) {
                $key = \bin2hex($c);
                if (!\array_key_exists($key, $idx)) {
                    continue;
                }
                $j = $idx[$key];
                if ($color[$j] === 1) {
                    return true;
                }
                if ($color[$j] === 0 && $hasCycle($j)) {
                    return true;
                }
            }
            $color[$i] = 2;
            return false;
        };

        for ($i = 0; $i < $count; $i++) {
            if ($color[$i] === 0 && $hasCycle($i)) {
                throw new CausalViolation("causal graph contains a cycle");
            }
        }
    }

    /**
     * Deterministic topological merge over the union causal DAG; ties break by content id.
     *
     * @param array<int,array{0:string,1:array<int,string>,2:int}> $nodes
     * @return array<int,string> ordered list of content ids (binary strings)
     */
    public static function reconcile(array $nodes): array
    {
        self::verifyCausal($nodes);
        $count = \count($nodes);
        $ids = [];
        $present = [];
        foreach ($nodes as $node) {
            $ids[] = $node[0];
            $present[\bin2hex($node[0])] = true;
        }
        $causes = [];
        foreach ($nodes as $node) {
            $kept = [];
            foreach ($node[1] as $c) {
                if (\array_key_exists(\bin2hex($c), $present)) {
                    $kept[] = $c;
                }
            }
            $causes[] = $kept;
        }
        $indeg = \array_map('count', $causes);
        $done = \array_fill(0, $count, false);
        $order = [];
        while (\count($order) < $count) {
            $pick = -1;
            for ($i = 0; $i < $count; $i++) {
                if ($done[$i] || $indeg[$i] !== 0) {
                    continue;
                }
                if ($pick === -1 || \strcmp($ids[$i], $ids[$pick]) < 0) {
                    $pick = $i;
                }
            }
            if ($pick === -1) {
                throw new CausalViolation("no ready node (unreachable after verifyCausal)");
            }
            $done[$pick] = true;
            $order[] = $ids[$pick];
            for ($j = 0; $j < $count; $j++) {
                if ($done[$j]) {
                    continue;
                }
                foreach ($causes[$j] as $c) {
                    if ($c === $ids[$pick]) {
                        $indeg[$j] -= 1;
                        break;
                    }
                }
            }
        }
        return $order;
    }

    /**
     * The tier-1 Reconcile body {1: [authorities], 2: [order content-ids]}.
     *
     * @param array<int,string> $authorities
     * @param array<int,string> $order binary content ids
     */
    public static function reconcileRecord(array $authorities, array $order): string
    {
        $auth = new A(\array_map(static fn(string $a): T => new T($a), $authorities));
        $ordr = new A(\array_map(static fn(string $o): B => new B($o), $order));
        return Cbor::encode(new M([
            [new U(1), $auth],
            [new U(2), $ordr],
        ]));
    }
}
